x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/AbstractMultivalueFunctionTestCase.java

```diff
@@ -7,129 +7,446 @@
+    ) {
+        cases.add(
+            new TestCaseSupplier(
+                name + "(\"\")",
+                () -> new TestCase(
+                    List.of(new TypedData(List.of(new BytesRef("")), DataTypes.KEYWORD, "field")),
+                    evaluatorName + "[field=Attribute[channel=0]]",
+                    expectedDataType,
+                    matcher.apply(1, Stream.of(new BytesRef("")))
+                )
+            )
+        );
+        cases.add(new TestCaseSupplier(name + "(BytesRef)", () -> {
+            BytesRef data = new BytesRef(randomAlphaOfLength(10));
+            return new TestCase(
+                List.of(new TypedData(List.of(data), DataTypes.KEYWORD, "field")),
+                evaluatorName + "[field=Attribute[channel=0]]",
+                expectedDataType,
+                matcher.apply(1, Stream.of(data))
+            );
+        }));
+        for (Block.MvOrdering ordering : Block.MvOrdering.values()) {
+            cases.add(new TestCaseSupplier(name + "(<BytesRefs>) " + ordering, () -> {
+                List<BytesRef> mvData = randomList(1, 100, () -> new BytesRef(randomAlphaOfLength(10)));
+                putInOrder(mvData, ordering);
+                return new TestCase(
+                    List.of(new TypedData(mvData, DataTypes.KEYWORD, "field")),
+                    evaluatorName + "[field=Attribute[channel=0]]",
+                    expectedDataType,
+                    matcher.apply(mvData.size(), mvData.stream())
+                );
+            }));
+        }
     }
 
-    @Override
-    protected DataType expectedType(List<DataType> argTypes) {
-        return argTypes.get(0);
+    /**
+     * Build a test case with {@code double} values.
+     */
+    protected static void doubles(
+        List<TestCaseSupplier> cases,
+        String name,
+        String evaluatorName,
+        BiFunction<Integer, DoubleStream, Matcher<Object>> matcher
+    ) {
+        doubles(cases, name, evaluatorName, DataTypes.DOUBLE, matcher);
     }
 
-    private Matcher<Object> resultsMatcher(List<TypedData> typedData) {
-        return resultMatcherForInput((List<?>) typedData.get(0).data(), typedData.get(0).type());
+    /**
+     * Build a test case with {@code double} values.
+     */
+    protected static void doubles(
+        List<TestCaseSupplier> cases,
+        String name,
+        String evaluatorName,
+        DataType expectedDataType,
+        BiFunction<Integer, DoubleStream, Matcher<Object>> matcher
+    ) {
+        cases.add(
+            new TestCaseSupplier(
+                name + "(0.0)",
+                () -> new TestCase(
+                    List.of(new TypedData(List.of(0.0), DataTypes.DOUBLE, "field")),
+                    evaluatorName + "[field=Attribute[channel=0]]",
+                    expectedDataType,
+                    matcher.apply(1, DoubleStream.of(0.0))
+                )
+            )
+        );
```

Comment:
On the original PR @not-napoleon mentioned this should also add the "fun" double values like NaN and -0 and stuff. There are like another 5 of these to add. I'll do that in a followup.